### PR TITLE
Implement remote parameter search engine client and BONES server

### PR DESCRIPTION
**Pull request by garymm**
Implement remote parameter search engine client and BONES server

Progress towards #202.

Still to do:

* Have jobczar actually make use of the remote engine.
* rename the optmize sub-directory and the files
  it contains to better match the names we're using in the Thrift files.
* Add unit tests for server and client. Current tests are integration
  tests.

### CODE REVIEW

**Comment by jedmccaleb** [comment]
idk about moving away from a simple string for the config values.  Maybe just do that on the remoteengine <-> bones side and keep it a string for all the other infra stuff

File: infra/jobczar/search.go [hunk 1]
```diff
@@ -18,23 +18,23 @@ import (
...
+	worldParam  *infra.ParamMap
+	agentParam  *infra.ParamMap
 	bufferSize  int // how many jobs to keep in the buffer
 	SearchSpace map[string]ParamRange
 	jobs        map[int32]int32 // map from jobID to pointID given by the search engine
 	engine      SearchEngine
 }
 
-func (search *Search) Init(cfgString string) {
+func (search *Search») Init(cfgString string) {
```
Comment (jedmccaleb) [inline infra/jobczar/search.go:37]: what are these little arrows?

**Comment by qemqemqem** [comment]
> idk about moving away from a simple string for the config values. Maybe just do that on the remoteengine <-> bones side and keep it a string for all the other infra stuff

@jedmccaleb Is the reason to keep it a string so that it has a canonical database representation? It seems like the natural representation is a dictionary, so we'll need to keep converting back and forth. But I know JobInfo has a correspondence with the database, so it's better for it to be a string there?

**Comment by jedmccaleb** [comment]
Ultimately we are writing it to a .cfg file on the machine that is doing the job and it is better for it to be human readable both there and in the DB. You guys don't have to convert back and forth in your dealing with bones. Just leave it as a dictionary for that but hand me back a string

File: infra/jobczar/search.go (hunk 1 above)
Comment (garymm) [reply infra/jobczar/search.go]: Not sure, but removed.

**Review by etuleu — APPROVED** [review]

File: infra/jobczar/remoteengine.go [hunk 2]
```diff
@@ -1,25 +1,73 @@
...
 
-func (engine *RemoteEngine) Init(search *Search) {
-	// TODO
+// Init configures the engine.
+func (engine *RemoteEngine) Init(search *Search) error {
+	engine.search = search
+	config := thrift.TConfiguration{
+		ConnectTimeout: 2*time.Second,
+		SocketTimeout:  2*time.Second}
+	// TODO: don't hard-code address
```
Comment (qemqemqem) [inline infra/jobczar/remoteengine.go:30]: TODO: TODO is done

File: infra/jobczar/search.go [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
 	log "github.com/Astera-org/easylog"
 	"github.com/Astera-org/obelisk/infra/gengo/infra"
 	"github.com/BurntSushi/toml"
 )
 
+type SearchEngine interface {
+	Init(search *Search) error
+	RecordResult(pointID int32, job *infra.JobInfo) error
+	GetNextJob() (point string, pointID int32, err error)
+	ShutDown()  // TODO: call this
```
Comment (qemqemqem) [inline infra/jobczar/search.go:14]: Clarify comment to say who needs to call it

File: infra/optimizer/bones_parameter_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
+from thrift.transport import TSocket
```
Comment (qemqemqem) [inline infra/optimizer/bones_parameter_optimizer.py:1]: Rename to bones_server.py or bones_suggester.py or something in subsequent PR.

File: infra/optimizer/optimize.thrift [hunk 5]
```diff
@@ -1,29 +1,26 @@
-namespace py optimize
-namespace go optimize
-namespace cpp optimize
+namespace py infra.optimize
+namespace go infra.optimize
 
-struct HyperParameter {
+include "../infra.thrift"
```
Comment (qemqemqem) [inline infra/optimizer/optimize.thrift:4]: Add a comment saying that this thrift is for communication between RemoteEngine and bones server, or something to that effect.

Comment (qemqemqem) [reply infra/optimizer/optimize.thrift:4]: Or move it to the other thrift file.

**Review by qemqemqem — APPROVED** [review]

**Pull request merged**